network/p2p/connection/connection_gater_test.go

```diff
@@ -0,0 +1,282 @@
+package connection_test
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/libp2p/go-libp2p/core/control"
+	"github.com/libp2p/go-libp2p/core/network"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/libp2p/go-libp2p/p2p/net/swarm"
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module/irrecoverable"
+	"github.com/onflow/flow-go/network/channels"
+	"github.com/onflow/flow-go/network/internal/p2pfixtures"
+	"github.com/onflow/flow-go/network/internal/testutils"
+	"github.com/onflow/flow-go/network/p2p"
+	mockp2p "github.com/onflow/flow-go/network/p2p/mock"
+	"github.com/onflow/flow-go/utils/unittest"
+)
+
+// TestConnectionGating tests node allow listing by peer ID.
+func TestConnectionGating(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	signalerCtx := irrecoverable.NewMockSignalerContext(t, ctx)
+
+	sporkID := unittest.IdentifierFixture()
+
+	// create 2 nodes
+	node1Peers := make(map[peer.ID]struct{})
+	node1, node1Id := p2pfixtures.NodeFixture(
+		t,
+		sporkID,
+		t.Name(),
+		p2pfixtures.WithConnectionGater(testutils.NewConnectionGater(func(p peer.ID) error {
+			if _, ok := node1Peers[p]; !ok {
+				return fmt.Errorf("id not found: %s", p.String())
+			}
+			return nil
+		})))
+
+	node2Peers := make(map[peer.ID]struct{})
+	node2, node2Id := p2pfixtures.NodeFixture(
+		t,
+		sporkID,
+		t.Name(),
+		p2pfixtures.WithConnectionGater(testutils.NewConnectionGater(func(p peer.ID) error {
+			if _, ok := node2Peers[p]; !ok {
+				return fmt.Errorf("id not found: %s", p.String())
+			}
+			return nil
+		})))
+
+	nodes := []p2p.LibP2PNode{node1, node2}
+	ids := flow.IdentityList{&node1Id, &node2Id}
+	p2pfixtures.StartNodes(t, signalerCtx, nodes, 100*time.Millisecond)
+	defer p2pfixtures.StopNodes(t, nodes, cancel, 100*time.Millisecond)
+
+	p2pfixtures.AddNodesToEachOthersPeerStore(t, nodes, ids)
+
+	t.Run("outbound connection to a disallowed node is rejected", func(t *testing.T) {
+		// although nodes have each other addresses, they are not in the allow-lists of each other.
+		// so they should not be able to connect to each other.
+		p2pfixtures.EnsureNoStreamCreationBetweenGroups(t, ctx, []p2p.LibP2PNode{node1}, []p2p.LibP2PNode{node2}, func(t *testing.T, err error) {
+			require.True(t, errors.Is(err, swarm.ErrGaterDisallowedConnection))
+		})
+	})
+
+	t.Run("inbound connection from an allowed node is rejected", func(t *testing.T) {
```

Comment:
I'm a little confused about the expected behavior based on the test title and comments. Can you elaborate a bit more?

Reply:
https://github.com/onflow/flow-go/pull/3459/commits/7e8a267e71e40daf05c9773c5fc6a46c942a8392